File: leetcode/Python/0305-h-number-of-islands-ii.py
"""
Union Find
"""
class Solution:
    def numIslands2(self, m: int, n: int, positions: List[List[int]]) -> List[int]:
        res = []
        island = set()
        self.size = 0
        self.parent = {}
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        for pos in positions:
            x, y = pos[0], pos[1]
            if (x, y) in island:
                res.append(self.size)
                continue
            
            island.add((x, y))
            self.parent[(x, y)] = (x, y)
            self.size += 1
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if (nx, ny) in island:
                    self.union((nx, ny), (x, y))
            res.append(self.size)
        return res
    
    def union(self, point_a, point_b):
        root_a = self.find(point_a)
        root_b = self.find(point_b)
        if root_a != root_b:
            self.parent[root_a] = root_b
            self.size -= 1
    
    def find(self, point):
        path = []
        while point != self.parent[point]:
            path.append(point)
            point = self.parent[point]

        for p in path:
            self.parent[p] = point

        return point

File: leetcode/C++/0305-h-number-of-islands-ii.cpp
class Solution {
private:
    int find_parent(int p, vector<int>& parent) {
        while (p != parent[p]) {
            parent[p] = parent[parent[p]];
            p = parent[p];
        }
        return p;
    }
    
    void union_op(int a, int b, int& islands, vector<int>& parent) {
        int parent_a = find_parent(a, parent);
        int parent_b = find_parent(b, parent);
        if (parent_a != parent_b) {
            parent[parent_a] = parent_b;
            islands--;
        }
    }
    
public:
    vector<int> numIslands2(int m, int n, vector<vector<int>>& positions) {
        vector<int> parent(m * n, -1);
        vector<vector<int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        vector<int> res;
        int islands = 0;
        for (int i = 0; i < positions.size(); i++) {
            int x = positions[i][0], y = positions[i][1];
            int idx_p = x * n + y;
            if (parent[idx_p] != -1) {
                res.push_back(islands);
                continue;
            }
            parent[idx_p] = idx_p;
            islands++;
            for (vector<int> d : directions) {
                int nx = x + d[0], ny = y + d[1];
                int idx_nei = nx * n + ny;
                if (0 <= nx && nx < m && 0 <= ny && ny < n && parent[idx_nei] != -1) {
                    union_op(idx_p, idx_nei, islands, parent);
                }
            }
            res.push_back(islands);
        }
        return res;
    }
};